### PR TITLE
ISPN-8860 Lazy start counter's caches

### PR DESCRIPTION
**Pull request by pruivo**
https://issues.jboss.org/browse/ISPN-8860

### CODE REVIEW

File: counter/src/main/java/org/infinispan/counter/impl/manager/CounterConfigurationManager.java [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+            if (configuration.concurrencyLevel() < 1) {
+               throw log.invalidConcurrencyLevel(configuration.concurrencyLevel());
+            }
+            break;
+      }
+   }
+
+   private void startCounterCache() {
+      if (counterCacheStarted.compareAndSet(false, true)) {
+         new Thread(() -> cacheManager.getCache(CounterModuleLifecycle.COUNTER_CACHE_NAME),
```
Comment (ryanemerson) [inline counter/src/main/java/org/infinispan/counter/impl/manager/CounterConfigurationManager.java]: Should we use one of the core executors here?

**Comment by pruivo** [comment]
@ryanemerson of course. I'm fetching the async operations executor to start the cache :+1:

**Comment by karesti** [comment]
@pruivo don't know if related but there are 2 counter failures 😢 
org.infinispan.counter.StrongCounterTest.testCompareAndSetConcurrent
org.infinispan.counter.RemoveCounterTest.testCounterManagerNonExistingRemoveWithBounded

**Comment by pruivo** [comment]
@karesti I'm aware. The fist is being tracked in another JIRA and the second should be fixed now. I have a third related to cluster listener that I need to fix :)

File: commons/src/main/java/org/infinispan/commons/util/Util.java [hunk 2]
```diff
@@ -1099,4 +1102,22 @@ public static String xmlify(String s) {
...
+         lock = lockFileOS.getChannel().lock();
+         if (!tempFile.renameTo(dstFile)) {
+            renameFailed.accept(tempFile, dstFile);
+         }
+      } finally {
+         if (lock != null && lock.isValid()) {
+            lock.release();
+         }
+         if (!lockFile.delete()) {
+            log.debugf("Unable to delete lock file %s", lockFile);
```
Comment (karesti) [inline commons/src/main/java/org/infinispan/commons/util/Util.java]: log debug here is in purpose ?

File: counter/src/main/java/org/infinispan/counter/configuration/ConvertUtil.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * @author Pedro Ruivo
+ * @since 9.2
+ */
+public final class ConvertUtil {
+
+   private ConvertUtil() {
+   }
+
+   public static CounterConfiguration parsedConfigToConfig(AbstractCounterConfiguration configuration) {
```
Comment (karesti) [inline counter/src/main/java/org/infinispan/counter/configuration/ConvertUtil.java]: I would prefer 3 methods and casting just once here
```
public static CounterConfiguration parsedConfigToConfig(AbstractCounterConfiguration configuration) {
      if (configuration instanceof StrongCounterConfiguration) {
         return parsedConfigToConfig((StrongCounterConfiguration) configuration);

      } else if (configuration instanceof WeakCounterConfiguration) {
         return parsedConfigToConfig((WeakCounterConfiguration) configuration);
      }

      throw new IllegalStateException(
            "[should never happen] unknown CounterConfiguration class: " + configuration.getClass());
   }
```

File: counter/src/main/java/org/infinispan/counter/logging/Log.java [hunk 4]
```diff
@@ -56,13 +54,13 @@
 //   @Message(value = "Unable to reset non-existing counter '%s'.", id = 28011)
 //   void noSuchCounterReset(ByteString counterName);
 
-   @LogMessage(level = WARN)
-   @Message(value = "Interrupted while waiting for the counter manager caches.", id = 28012)
-   void interruptedWhileWaitingForCaches();
+//   @LogMessage(level = WARN)
```
Comment (karesti) [inline counter/src/main/java/org/infinispan/counter/logging/Log.java]: did you leave these lines commented for any reason ?

File: counter/src/test/java/org/infinispan/counter/AbstractCounterNotificationTest.java [hunk 5]
```diff
@@ -169,32 +171,41 @@ private void printQueue(ListenerQueue queue) {
...
          queue = new LinkedBlockingQueue<>();
+         this.name = name;
       }
 
       @Override
       public void onUpdate(CounterEvent event) {
-         queue.offer(event);
+         try {
+            queue.put(event);
+            log.tracef("[%s] add event %s", name, event);
```
Comment (karesti) [inline counter/src/test/java/org/infinispan/counter/AbstractCounterNotificationTest.java]: isTraceEnabled is not necessary here ?

File: counter/src/main/java/org/infinispan/counter/logging/Log.java (hunk 4 above)
Comment (pruivo) [reply counter/src/main/java/org/infinispan/counter/logging/Log.java]: yes. we started to comment unused log messages per https://issues.jboss.org/browse/ISPN-7456

File: commons/src/main/java/org/infinispan/commons/util/Util.java (hunk 2 above)
Comment (pruivo) [reply commons/src/main/java/org/infinispan/commons/util/Util.java]: yes, just to remove the yellow background warning that I'm ignoring the `delete()` result.

File: counter/src/test/java/org/infinispan/counter/AbstractCounterNotificationTest.java (hunk 5 above)
Comment (pruivo) [reply counter/src/test/java/org/infinispan/counter/AbstractCounterNotificationTest.java]: not really, because it is a test. but I'll add it.

Comment (karesti) [reply counter/src/test/java/org/infinispan/counter/AbstractCounterNotificationTest.java]: ah I didn't realise the name of the class!

File: counter/src/main/java/org/infinispan/counter/logging/Log.java (hunk 4 above)
Comment (karesti) [reply counter/src/main/java/org/infinispan/counter/logging/Log.java]: ok

File: commons/src/main/java/org/infinispan/commons/util/Util.java (hunk 2 above)
Comment (karesti) [reply commons/src/main/java/org/infinispan/commons/util/Util.java]: ok

**Comment by karesti** [comment]
@pruivo let me know when is good from you side and I merge!

**Comment by pruivo** [comment]
@karesti let CI run first :)

**Comment by karesti** [comment]
org.infinispan.counter.WeakCounterNotificationTest.testMultipleListeners failed but another issue has been merged that should correct it.
Let's rebase, let CI pass and merge! 💃

**Pull request merged**